### PR TITLE
Additional Calculations & Roll, Up, Down Helpers

### LINKED ISSUE #41: Remove Information Ratio
Our Information Ratio calculation (https://github.com/quantopian/empyrical/blob/master/empyrical/stats.py#L590) is wrong as noted by @marketneutral:
"The definition is wrong. You need a risk model to calc active risk to benchmark. And the benchmark look through to calc active return."

As this is still a ways off, we should rather delete it instead of having a wrong measure in here.

### PR DESCRIPTION
**Pull request by cgdeboer**
## Summary
Added a few helper methods in the utils file that allow us to stack some transformation logic together before calling an actual metric. For example, if we want to calculate the rolling up capture of a strategy vs its benchmark, we want to be able to do something like `[roll,up, capture]` where `roll`, `up`, `capture`, are each methods that perform either a series deconstruction (roll), a filter (up), or a calculation (capture). Doing it this why is more DRY than embedding roll and filter logic in each method, but does add some complication. 

I elected to add a variety of small methods to the `stats` file that wrap the complexity described above. This way we keep the API the user is used to calling the same. e.g there is actually a method called `rolling_up_capture` that takes the expected "returns, factor_returns" arguments. That function simply calls the more complex stacked function to get the results.

## Details
- added new methods details below.
- added roll, up, and down as stackable methods to use in future calculations
- added a handful of tests for new methods
- corrected an issue with the `max_drawdown`. See note below
- updated `nose_parameterized` to just `parameterized`
- move date/period symbols into a separate file
- added some examples to the README

## Not Done
- add `information_ratio` per #41 
- bump `empyrical` version
- add remainder of tests.

#### Max Drawdown
The `max_drawdown` code was not considering the first return in the array. The `max_return` array did not consider the starting value of 100 used in the function. In cases where ALL the returns in the array are negative, the max value in the array will be the 100 * (1+ the first negative return). For e.g if the first return in the array is `-0.02`, then the max value in the array will be 98. This will cause max drawdowns to appear less severe than if they started at 100. 

The max drawdown code was masquerading as correct because the `negative_returns` Series in the test suite starts with a 0. (i.e they aren't all negative). 

I've corrected that by ensuring that the `cum_returns` array starts with a value that matches the base value.

#### New Methods
```
    capture,
    up_capture,
    down_capture,
    up_down_capture,
    up_alpha_beta,
    down_alpha_beta,
    roll_max_drawdown,
    roll_up_capture,
    roll_down_capture,
    roll_up_down_capture,
    roll_alpha_beta,
    roll_sharpe_ratio,
```

Resolves #41 #46

### CODE REVIEW

File: empyrical/stats.py [hunk 1]
```diff
@@ -587,7 +572,7 @@ def downside_risk(returns, required_return=0, period=DAILY,
     return dside_risk
 
 
-def information_ratio(returns, factor_returns):
+def excess_sharpe(returns, factor_returns):
     """
     Determines the Information ratio of a strategy.
```
Comment (twiecki) [inline empyrical/stats.py]: need to change doc string.

File: empyrical/stats.py [hunk 2]
```diff
@@ -1004,6 +990,208 @@ def cagr(returns, period=DAILY, annualization=None):
 
     return ending_value ** (1. / no_years) - 1
 
+def capture(returns, factor_returns, period=DAILY):
+    """
```
Comment (twiecki) [inline empyrical/stats.py]: needs a reference

**Comment by twiecki** [comment]
@cgdeboer I won't do a full review yet as you label it as WIP but wanted to say that this looks fantastic.

**Comment by cgdeboer** [comment]
@twiecki I'll get this thing linted etc, and address your initial comments above. I'll remove the WIP title and notify you here once that's done.

File: empyrical/stats.py (hunk 2 above)
Comment (cgdeboer) [inline empyrical/stats.py]: What kind of reference ? 

- http://www.morningstar.com/InvGlossary/upside-downside-capture-ratio.aspx

- http://www.investopedia.com/terms/u/up-market-capture-ratio.asp

Something like that? or am I misunderstanding what you mean by reference ?

**Comment by cgdeboer** [comment]
@twiecki alrighty, I still have some tests to write and verify, and I'm still conferring with some colleagues on the info ratio stuff (i.e how to best proceed with that). At my own peril, I'm removing the WIP, review away.

**Comment by cgdeboer** [comment]
@twiecki et al, let me know what you think. Would love to get this merged in this week or next.

**Comment by twiecki** [comment]
@cgdeboer will take a look soon!

File: setup.py [hunk 3]
```diff
@@ -61,7 +61,6 @@
     'numpy>=1.9.2',
     'pandas>=0.16.1',
     'scipy>=0.15.1',
-    'bottleneck>=1.0.0'
```
Comment (twiecki) [inline setup.py]: Why?

File: empyrical/stats.py [hunk 4]
```diff
@@ -720,11 +708,11 @@ def alpha_beta_aligned(returns, factor_returns, risk_free=0.0, period=DAILY,
...
     Returns
     -------
-    float
-        Alpha.
-    float
-        Beta.
-
+    list
+        float
+            Alpha.
```
Comment (twiecki) [inline empyrical/stats.py]: mixed tabs and spaces?

File: empyrical/stats.py [hunk 5]
```diff
@@ -1005,6 +993,214 @@ def cagr(returns, period=DAILY, annualization=None):
...
+        annualizing. Value ignored if `annualization` parameter is specified.
+        Defaults are:
+            'monthly':12
+            'weekly': 52
+            'daily': 252
+    Returns
+    -------
+    float, np.nan
+        The capture ratio.
+    """
```
Comment (twiecki) [inline empyrical/stats.py]: Can you add links to http://www.investopedia.com/terms/u/up-market-capture-ratio.asp ?

File: empyrical/stats.py [hunk 6]
```diff
@@ -1018,11 +1214,12 @@ def cagr(returns, period=DAILY, annualization=None):
     stats.skew,
     stats.kurtosis,
     tail_ratio,
-    cagr
+    cagr,
+    capture
```
Comment (twiecki) [inline empyrical/stats.py]: why not add up and down captures here?

File: empyrical/utils.py [hunk 7]
```diff
@@ -24,12 +26,63 @@
...
     nanmean = np.nanmean
     nanstd = np.nanstd
     nansum = np.nansum
     nanmax = np.nanmax
     nanmin = np.nanmin
     nanargmax = np.nanargmax
     nanargmin = np.nanargmin
+
+
+def roll(*args, **kwargs):
```
Comment (twiecki) [inline empyrical/utils.py]: all these need doc-strings.

File: empyrical/utils.py [hunk 8]
```diff
@@ -24,12 +26,63 @@
...
+
+def _roll_ndarray(func, window, *args, **kwargs):
+    data = []
+    for i in range(window, len(args[0])):
+        rets = [s[i-window:i] for s in args]
+        data.append(func(*rets, **kwargs))
+    return np.array(data)
+
+
+def _roll_pandas(func, window, *args, **kwargs):
```
Comment (twiecki) [inline empyrical/utils.py]: these helper functions definitely need tests, easy for edge cases / errors to creep in.

File: empyrical/stats.py [hunk 9]
```diff
@@ -1005,6 +993,214 @@ def cagr(returns, period=DAILY, annualization=None):
...
+
+    Parameters
+    ----------
+    see documentation for `capture` (pass all args, kwargs required)
+
+    window : int, required
+        Size of the rolling window in terms of the periodicity of the data.
+        - eg window = 60, periodicity=DAILY, represents a rolling 60 day window
+    """
+    return roll(returns, factor_returns, window=window, functions=[
```
Comment (twiecki) [inline empyrical/stats.py]: couldn't you pass `functions=[up_capture]` instead? If functions is ever only a single argument might make the code below easier too. Or a lambda could be used here, or something neat from functools or itertools.

**Comment by twiecki** [comment]
@cgdeboer This looks great, many thanks for the refactor / contribution. I left some comments, mainly tests for the utils functions would be good. But generally I find the way you coded things up very elegant. Let me know if you have any questions.

**Comment by cgdeboer** [comment]
@twiecki I'll address your comments by the end of this week. Thanks for the look !

File: setup.py (hunk 3 above)
Comment (cgdeboer) [inline setup.py]: From `try:except` block in the `utils.py` file it looks like the use of the bottleneck library is intended to be optional (for those who need the bone rattlin' speed of optimized C ).  Removing bottleneck from the hard requirements is merely a nod to the logic in the `utils.py` file. 

It seems like we might want to either: 

1. remove bottleneck from the requirements (as I have done), or 
2. simplify the utils file to always use the bottleneck library. Either seem logical to me. 

@twiecki 's choice 😃 

(if there is some other reason to keep it as it originally was, I'm happy to revert to that as well... perhaps some sort of thing where `bottleneck` is only supported in certain architectures? )

Comment (richafrank) [inline setup.py]: I think bottleneck needs to be required at the moment. We [discovered](https://github.com/quantopian/empyrical/pull/45#issuecomment-299338946) that it has diverged slightly from numpy, and we rely on the bottleneck behavior.

File: empyrical/stats.py (hunk 6 above)
Comment (cgdeboer) [inline empyrical/stats.py]: Added em in a89510e . I also realized I added capture to the wrong list, it belongs to FACTOR_STAT_FUNCS, so I put capture (reg, up and down) all in there.

File: empyrical/stats.py (hunk 5 above)
Comment (cgdeboer) [inline empyrical/stats.py]: Added in a89510e

File: empyrical/stats.py (hunk 4 above)
Comment (cgdeboer) [inline empyrical/stats.py]: Not sure how that happened (the tab in there). Also, I don't think I intended to keep the doc string "Returns" syntax with the `list` nomenclature, that was some WIP I had in there to remind me that a list came back with that. I've reverted this to the original "pair o' floats" style "Returns" section, and ensured that's how it is across the board.

File: setup.py (hunk 3 above)
Comment (cgdeboer) [inline setup.py]: noted.

Comment (cgdeboer) [inline setup.py]: Fixed in aa07b5c

File: empyrical/utils.py (hunk 7 above)
Comment (cgdeboer) [inline empyrical/utils.py]: Fixed in 6049309

File: empyrical/utils.py (hunk 8 above)
Comment (cgdeboer) [inline empyrical/utils.py]: added in 	ab1eabe

File: empyrical/stats.py (hunk 9 above)
Comment (cgdeboer) [inline empyrical/stats.py]: I think your idea about a single function makes the most sense. The `functions=[]` was sort of over-engineered. I changed it all to take a `function` kwarg in 	71ea444

**Comment by cgdeboer** [comment]
@twiecki I think all of your comments have been addressed. Looks like ole Travis is just finishing up the CI testing (or maybe hung up on one build job).

File: empyrical/stats.py [hunk 10]
```diff
@@ -1005,6 +992,227 @@ def cagr(returns, period=DAILY, annualization=None):
...
+
+    Parameters
+    ----------
+    see documentation for `capture` (pass all args, kwargs required)
+
+    window : int, required
+        Size of the rolling window in terms of the periodicity of the data.
+        - eg window = 60, periodicity=DAILY, represents a rolling 60 day window
+    """
+    return roll(returns, factor_returns, window=window, functions=up_capture,
```
Comment (twiecki) [inline empyrical/stats.py]: function

Comment (cgdeboer) [inline empyrical/stats.py]: Newman !! Fixed it in 6efeee1

**Comment by cgdeboer** [comment]
Alrighty @twiecki , addressed that last pluralized one. Let me know if you see anything else, could catch.

**Comment by twiecki** [comment]
Did the unittest catch it as well?

**Comment by cgdeboer** [comment]
Well. You know what, I don't think it did. Perhaps I need to write a functional test for each of the new methods. I sort of thought we'd rely on the individual stat tests for "financial" accuracy, and a sample of tests for the wrappers like `up`, `down`, and `roll`.  

I could see it going either way. What is your preference?

**Comment by twiecki** [comment]
Best to have it run through all code paths.

**Comment by cgdeboer** [comment]
Will add tests for all the methods. Back in a few.

**Comment by twiecki** [comment]
@cgdeboer Any update?

**Comment by cgdeboer** [comment]
@twiecki should have em all done this week. (My few referenced above turned from a few hours to a few days). Other stuff hopped in the way. Shooting for Thursday morning.

**Comment by twiecki** [comment]
OK, no rush obviously, just wondered if we could help.

**Comment by cgdeboer** [comment]
Most of the new tests are now written, just taking some extra time to double check the inputs vs expected.

**Comment by cgdeboer** [comment]
@twiecki tests added. 🏀  is in your court.

File: empyrical/utils.py [hunk 11]
```diff
@@ -12,24 +12,125 @@
...
-    # slower numpy
-    import numpy as np
-    nanmean = np.nanmean
-    nanstd = np.nanstd
-    nansum = np.nansum
-    nanmax = np.nanmax
-    nanmin = np.nanmin
-    nanargmax = np.nanargmax
-    nanargmin = np.nanargmin
+nanmean = bn.nanmean
```
Comment (twiecki) [inline empyrical/utils.py]: This should be reverted.

Comment (cgdeboer) [inline empyrical/utils.py]: Reverted in dfe495f.

**Comment by twiecki** [comment]
Great work, thanks @cgdeboer!

**Comment by cgdeboer** [comment]
Thanks !. @twiecki  I'll keep an eye on https://pypi.python.org/pypi/empyrical/ for version update ?

**Comment by twiecki** [comment]
@cgdeboer New release should be up.
